engine/crates/partial-caching/src/planning/fragments/tracker.rs

```diff
@@ -26,10 +60,24 @@ impl super::super::visitor::Visitor for FragmentTracker {
         self.selection_stack.pop();
     }
 
+    fn enter_field(&mut self, edge: FieldEdge<'_>) {
+        if let Some(cache_control) = edge.cache_control() {
+            self.cache_control_stack.push((*cache_control).clone());
+        }
+    }
+
+    fn exit_field(&mut self, edge: FieldEdge<'_>) {
+        if edge.cache_control().is_some() {
+            self.cache_control_stack.pop();
+        }
+    }
+
     fn fragment_spread(&mut self, spread: FragmentSpread<'_>) {
-        self.used_fragments
-            .entry(spread.fragment_name().to_string())
-            .or_default()
-            .extend(self.selection_stack.iter().copied())
+        let Some(key) = self.key_for(spread) else {
```

Comment:
why not raise an error here? instead of doing it later?

Reply:
Because this is the implementation of a trait and the trait function doesn't have a return type to raise an error through.

I could add one, but I think that would make the trait & visitor code worse.